googleapis/sdk-platform-java
# feat: add client side logging with slf4j

Changes to introduce client logging debug feature to Gax. Guide on usage will be added separately to README file.

- Brings in `slf4j-api` as optional dependency to gax-java
- `LoggingUtils`, `LogData` and `LoggerProvider` are public because access are needed from gax-grpc and gax-httpjson packages
  - `LoggingUtils` handles logic to enable/disable from env var, and contains shared utility methods for record data for logging and record logs (abstracting away any need for SLF4J classes). 
  - `Slf4jUtils` any logic interacting with SLF4J classes.
  - `LoggerProvider` provides the SLF4J Logger. This is so that Logger interceptor classes do not declare Logger directly.
- This feature is guarded by env var GOOGLE_SDK_JAVA_LOGGING, only turned on if true. 
- By default it is off, user app should act as before (usual tests behaves as usual)

__Tests added:__
- Unit tests in Gax that need either GOOGLE_SDK_JAVA_LOGGING unset, or do not depend on env var
- LoggingEnabledTest: test for logger correctly setups when GOOGLE_SDK_JAVA_LOGGING = true. This is added to existing `envVarTest` profile
- Showcase tests:
  - ITLoggingDisabled: test no logging event should record when env var is turned off.
  - ITLogging: test logging event with slf4j2.x+logback
  - ITLogging1x: test logging event with slf4j1.x+logback
These tests may not compile depending on the logging dependency brought in. Set up profiles to control test compile and run:
    - Slf4j2_logback, slf4j1_logback, disabledLogging: brings in logging dependencies and setup compile config exclusions
    - Do not include `it/logging` folder for compile by default, or `enable-golden-tests`. `native` is also excluded for now.

__Notable changes since last reviewed:__
- Changes related to get slf4j-api deps as optional
- Logics to switch logging behavior based on slf4j major version
- Moved helper methods into Gax
- Revamped test setup
- use Protobuf utils to serialize message

__Context:__ [go/java-client-logging-design](http://goto.google.com/java-client-logging-design)

TODO:
- add rules for renovate bot to bypass some versions for testing
